pkg/agent/ipassigner/ip_assigner_linux.go

```diff
@@ -148,28 +163,75 @@ func (a *ipAssigner) UnassignIP(ip string) error {
 		return fmt.Errorf("invalid IP %s", ip)
 	}
 
-	a.mutex.Lock()
-	defer a.mutex.Unlock()
-
-	if !a.assignedIPs.Has(ip) {
+	if !a.isIPAssigned(ip) {
 		klog.V(2).InfoS("The IP is not assigned", "ip", ip)
 		return nil
 	}
 
-	addr := util.NewIPNet(parsedIP)
-	if err := netlink.AddrDel(a.dummyDevice, &netlink.Addr{IPNet: addr}); err != nil {
-		return fmt.Errorf("failed to delete IP %v from interface %s: %v", ip, a.dummyDevice.Attrs().Name, err)
+	if a.dummyDevice != nil {
+		addr := util.NewIPNet(parsedIP)
+		if err := netlink.AddrDel(a.dummyDevice, &netlink.Addr{IPNet: addr}); err != nil {
+			if !errors.Is(err, unix.EADDRNOTAVAIL) {
+				return fmt.Errorf("failed to delete IP %v from interface %s: %v", ip, a.dummyDevice.Attrs().Name, err)
+			} else {
+				klog.InfoS("IP does not exist on interface", "ip", parsedIP, "interface", a.dummyDevice.Attrs().Name)
+			}
+		}
+		klog.InfoS("Deleted IP from interface", "ip", ip, "interface", a.dummyDevice.Attrs().Name)
 	}
-	klog.InfoS("Deleted IP from interface", "ip", ip, "interface", a.dummyDevice.Attrs().Name)
 
-	a.assignedIPs.Delete(ip)
+	if utilnet.IsIPv4(parsedIP) && a.arpResponder != nil {
+		if err := a.arpResponder.RemoveIP(parsedIP); err != nil {
+			return fmt.Errorf("failed to remove IP %v from ARP responder: %v", ip, err)
+		}
+	}
+	if utilnet.IsIPv6(parsedIP) && a.ndpResponder != nil {
+		if err := a.ndpResponder.RemoveIP(parsedIP); err != nil {
+			return fmt.Errorf("failed to remove IP %v from NDP responder: %v", ip, err)
+		}
+	}
+
+	a.deleteIP(ip)
 	return nil
 }
 
+func (a *ipAssigner) isIPAssigned(ip string) bool {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	return a.assignedIPs.Has(ip)
+}
+
+func (a *ipAssigner) deleteIP(ip string) {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	a.assignedIPs.Delete(ip)
+}
+
+func (a *ipAssigner) addIP(ip string) {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	a.assignedIPs.Insert(ip)
+}
+
 // AssignedIPs return the IPs that are assigned to the dummy device.
 func (a *ipAssigner) AssignedIPs() sets.String {
 	a.mutex.RLock()
 	defer a.mutex.RUnlock()
 	// Return a copy.
 	return a.assignedIPs.Union(nil)
 }
+
+// Run starts the ARP responder and NDP responder.
+func (a *ipAssigner) Run(ch <-chan struct{}) {
+	// Start the ARP responder only when the dummy device is not created. The kernel will handle ARP requests
+	// for IPs assigned to the dummy devices by default.
+	// TODO: Check arp_ignore sysctl parameter of the transport interface to determin whether to start
```

Comment:
determin -> determine

the arp_ignore sysctl parameter

Reply:
thanks!